### PR TITLE
api, server: fix add-remove vpn user without vpn owner

### LINKED ISSUE #5711: Error while create VPN user for another account
##### ISSUE TYPE
 * Bug Report

##### COMPONENT NAME
~~~
UI?
~~~

##### CLOUDSTACK VERSION
~~~
4.16.0
~~~

##### CONFIGURATION
XCP-NG 8.2
Advanced Zone

##### OS / ENVIRONMENT
Oracle Linux 8.4 - Management server, MYSQL 8.0.26
XCP-NG 8.2 - hypervisor

##### SUMMARY
Error while create VPN user for another account:

image

image
 
but user was added:

image

and when U delete it U wil get error:

image

##### STEPS TO REPRODUCE
~~~
Create new Isolated network. Enable VPN. Try to add new VPN user.
~~~
##### EXPECTED RESULTS
~~~
Can create VPN user. Can create VPN user for another account. 
~~~

##### ACTUAL RESULTS
~~~
Can create VPN user (if U dont specify domain and account). Can't create VPN user for another account (if U specify Doamin and Account). 
~~~
After we "creat" user in DB we can see:

image

After we "remove" user:

image

Only delete from database allow create same named VPN user.

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #5711

- When the owner account does not have VPN access, VPN user should be added in Add state
- While removing VPN user when the owner account has no VPN, VPN user should be deleted from the database.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@shwstppr 
I tested this pr just now. 
I noticed your comment (https://github.com/apache/cloudstack/issues/5711#issuecomment-1009803647), however I think users should be able to create vpn user even if there is no remote access vpn enabled.

the issues I found are
(1) cannot create vpn user for other account which do not have any isolated network.
(2) exception when create vpn user for other account which has isolated network but in 'Allocated' state. but vpn user is added in DB.
(2) exception when remove vpn user for other account which has isolated network but in 'Allocated' state. but vpn user is removed in DB.

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@weizhouapache thanks for testing. Will check and make the fixes

**Review by borisstoyanov — APPROVED** [review]
LGTM, manually tested

**Comment by shwstppr** [comment]
> @shwstppr I tested this pr just now. I noticed your comment ([#5711 (comment)](https://github.com/apache/cloudstack/issues/5711#issuecomment-1009803647)), however I think users should be able to create vpn user even if there is no remote access vpn enabled.
> 
> the issues I found are (1) cannot create vpn user for other account which do not have any isolated network. (2) exception when create vpn user for other account which has isolated network but in 'Allocated' state. but vpn user is added in DB. (2) exception when remove vpn user for other account which has isolated network but in 'Allocated' state. but vpn user is removed in DB.

@weizhouapache 
For adding user for an account without VPN access, currently code searches for available VPNs. Should we simply skip the apply part and just create the user in this case?
1. Should we allow creating user in this case?
2. Again, should we allow creating user in this case or should we show a proper error message?
3. User should get deleted in this case without error?

**Comment by weizhouapache** [comment]
> > @shwstppr I tested this pr just now. I noticed your comment ([#5711 (comment)](https://github.com/apache/cloudstack/issues/5711#issuecomment-1009803647)), however I think users should be able to create vpn user even if there is no remote access vpn enabled.
> > the issues I found are (1) cannot create vpn user for other account which do not have any isolated network. (2) exception when create vpn user for other account which has isolated network but in 'Allocated' state. but vpn user is added in DB. (2) exception when remove vpn user for other account which has isolated network but in 'Allocated' state. but vpn user is removed in DB.
> 
> @weizhouapache For adding user for an account without VPN access, currently code searches for available VPNs. Should we simply skip the apply part and just create the user in this case?
> 
> 1. Should we allow creating user in this case?
> 2. Again, should we allow creating user in this case or should we show a proper error message?
> 3. User should get deleted in this case without error?

@shwstppr 
VPN users belong to account, not to networks. I think we should allow creating/removing VPN users no matter what the networks' state is.

(btw, I think VPN users should be associated to networks, not account. but it looks like a big change in DB, service layer and UI).

**Comment by DaanHoogland** [comment]
> (btw, I think VPN users should be associated to networks, not account. but it looks like a big change in DB, service layer and UI).

@weizhouapache couldn't several networks use the same vpn user? I am using my vpn account from several locations, but it is exclusively mine. So as a parallel in cloudstack my account should own it.

that said is this maybe too big a discussion to plan anything for 4.16.1? cc @shwstppr @sureshanaparti

**Comment by weizhouapache** [comment]
> > (btw, I think VPN users should be associated to networks, not account. but it looks like a big change in DB, service layer and UI).
> 
> @weizhouapache couldn't several networks use the same vpn user? I am using my vpn account from several locations, but it is exclusively mine. So as a parallel in cloudstack my account should own it.
> 
> that said is this maybe too big a discussion to plan anything for 4.16.1? cc @shwstppr @sureshanaparti

@DaanHoogland 
if vpn user is owned by network, we can still add same vpn user to all networks , just a bit more steps.
if vpn user is owned by account, all networks will have same vpn users. it is not possible to add vpn user for specific network.

**Review by rohityadavcloud — APPROVED** [review]

**Comment by rohityadavcloud** [comment]
(my LGTM is to unblocking for merging) Only merge this after review/discuss and when smoketest pass

**Comment by DaanHoogland** [comment]
> @DaanHoogland
> if vpn user is owned by network, we can still add same vpn user to all networks , just a bit more steps.
> if vpn user is owned by account, all networks will have same vpn users. it is not possible to add vpn user for specific network.

Ok @weizhouapache, the real solution would then be to allow multiple vpn users per account and let the user choose one on opening a vpn-gateway, would it?

**Comment by weizhouapache** [comment]
> > @DaanHoogland
> > if vpn user is owned by network, we can still add same vpn user to all networks , just a bit more steps.
> > if vpn user is owned by account, all networks will have same vpn users. it is not possible to add vpn user for specific network.
> 
> Ok @weizhouapache, the real solution would then be to allow multiple vpn users per account and let the user choose one on opening a vpn-gateway, would it?

yes, that's an option. it is similar as sshkey or affinitygroup to vm

another option, network have its own vpn users, similar as volumes to vm, private gateway to vpc.

both require api,db,ui changes.

**Comment by shwstppr** [comment]
@weizhouapache @DaanHoogland @sureshanaparti cc @borisstoyanov I've made the change to fix behaviour when the user's owner account doesn't have RemoteVPNAccess (either no network with VPN access or the network is not in `Implemented` state).
add/delete API will simply add/delete entries from DB in these cases.
Maybe for better handling of VPN users we can have an improvement issue for future release

@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@shwstppr , these ^^ errors seem to be consistent. Can you have a look?

**Comment by shwstppr** [comment]
Test error seems to be related to env/host issue:
`{errorcode : 431, errortext : 'Cannot migrate VM, destination host is not in correct state, has status: Up, state: Disabled'}`

Trying tests again
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by weizhouapache — APPROVED** [review]
I have tested some scenarios, the following are good.

create vpn users for other accounts (I believe same results when create vpn users for same account by api)
(1) has no isolated network
(2) has isolated network in Allocated state, and no remote access vpn
(3) has isolated network in Implemented state, and no remote access vpn
(4) has isolated network in Implemented state, and remote access vpn is enabled
(5) has isolated network in Allocated state, and remote access vpn is enabled

I did not test
(1) has multiple isolated networks, but fail to apply vpn user in one of the VRs.

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by sureshanaparti** [comment]
> I have tested some scenarios, the following are good.
> 
> create vpn users for other accounts (I believe same results when create vpn users for same account by api) (1) has no isolated network (2) has isolated network in Allocated state, and no remote access vpn (3) has isolated network in Implemented state, and no remote access vpn (4) has isolated network in Implemented state, and remote access vpn is enabled (5) has isolated network in Allocated state, and remote access vpn is enabled
> 
> I did not test (1) has multiple isolated networks, but fail to apply vpn user in one of the VRs.

Hi @shwstppr @weizhouapache what's the concensus here, is this PR good to go / merge?

**Comment by weizhouapache** [comment]
> > I have tested some scenarios, the following are good.
> > create vpn users for other accounts (I believe same results when create vpn users for same account by api) (1) has no isolated network (2) has isolated network in Allocated state, and no remote access vpn (3) has isolated network in Implemented state, and no remote access vpn (4) has isolated network in Implemented state, and remote access vpn is enabled (5) has isolated network in Allocated state, and remote access vpn is enabled
> > I did not test (1) has multiple isolated networks, but fail to apply vpn user in one of the VRs.
> 
> Hi @shwstppr @weizhouapache what's the concensus here, is this PR good to go / merge?

@sureshanaparti 
it is good to me. I have approved it.

**Review by borisstoyanov — APPROVED** [review]
LGTM, manually tested

File: api/src/main/java/com/cloud/network/vpn/RemoteAccessVpnService.java [hunk 1]
```diff
@@ -43,6 +43,10 @@
 
     List<? extends VpnUser> listVpnUsers(long vpnOwnerId, String userName);
 
+    void removeVpnUserWithoutValidVpnOwner(long id);
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply api/src/main/java/com/cloud/network/vpn/RemoteAccessVpnService.java:46]: @weizhouapache yes, when VPN user needs to be just removed from DB https://github.com/apache/cloudstack/pull/5850/files#diff-910f0c63c2b80a2ac89ea6abd67c6c16238e819348f0388f84e939efb3162d28R551

[the remark this answers is not in the document]
Comment (weizhouapache) [reply api/src/main/java/com/cloud/network/vpn/RemoteAccessVpnService.java:46]: @shwstppr it uses another method `removeVpnUserWithoutRemoteAccessVpn`

[the remark this answers is not in the document]
Comment (shwstppr) [reply api/src/main/java/com/cloud/network/vpn/RemoteAccessVpnService.java]: @weizhouapache sorry realized that. I've removed the unused method now

[the remark this answers is not in the document]
Comment (weizhouapache) [reply api/src/main/java/com/cloud/network/vpn/RemoteAccessVpnService.java]: @shwstppr good !

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
> Trillian test result (tid-3257) Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7 Total time taken: 30690 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5850-t3257-kvm-centos7.zip Smoke tests completed. 91 look OK, 1 have errors Only failed tests results shown below:
> 
> Test	Result	Time (s)	Test File
> test_01_nic	`Error`	145.22	test_nic.py

Not noticed this test failure in the last run, seems to be  intermittent and is unrelated to this PR changes.

**Pull request merged**